test/release/examples/primers/forallLoops.chpl

```diff
@@ -0,0 +1,305 @@
+// Forall Loops
+
+/*
+This primer illustrates forall loops, which are a way to leverage
+data parallelism or engage :ref:`user-defined parallel iterators
+<primers-parIters>`. 
+
+Like serial for-loops, forall loops can iterate over a data structure,
+an iterator, or a zippered combination of these. Unlike for-loops,
+multiple iterations of a forall-loop can potentially execute in parallel.
+Parallelism is determined by the data structure or iterator being iterated
+over. For a zippered forall loop, parallelism is determined by the "leader",
+which is the first data structure or iterator in the zippered list.
+
+Chapel has forall statements and forall expressions. Each form has
+two varieties, "must-parallel" and "may-parallel".
+
+ - The must-parallel forms are written using the ``forall`` keyword.
+   They require that the leader iterable provide a parallel iterator.
+   Note that there are no requirements on the behavior of the parallel
+   iterator. For example, it can execute serially, in which case
+   the "must-parallel" loop that invokes it also executes serially.
+
+ - The may-parallel forms are written using brackets ``[ ]``.
+   They invoke the parallel iterator if the leader provides it,
+   otherwise fall back on the serial iterator.
+
+As with for-loops, the body of a forall statement is a statement or a
+a block statement, whereas the body of a forall expression is an expression.
+
+"Must-parallel" forall statement
+--------------------------------
+
+In the following example, the forall loop iterates over the array indices
+in parallel:
+*/
+
+config const n = 9;
+var A: [1..n] real;
+
+forall i in 1..n {
+  A[i] = i;
+}
+
+writeln("After setting up, A is:");
+writeln(A);
+writeln();
+
+/*
+If ``A`` were a distributed array, each loop iteration would be
+executed on the locale where the corresponding array element resides.
+
+"Must-parallel" forall expression
+---------------------------------
+
+The following forall-expression produces new values in parallel.
+We store these values in a new array.
+*/
+
+var B = forall a in A do a * 3;
+
+writeln("After initialization, B is:");
+writeln(B);
+writeln();
+
+/*
+Zippered "must-parallel" forall statement
+-----------------------------------------
+
+We can iterate over multiple arrays in parallel with a zippered loop.
+Here we illustrate zippering arrays and domains:
+*/
+
+var C: [1..n] real;
+forall (a, b, i) in zip(A, B, C.domain) do
+  C[i] = a * 10 + b / 10;
```

Comment:
This example's a little goofy since `i` isn't used for anything other than indexing into C.  I.e., if a user were actually writing this, we'd advise them to iterate over C directly as with A and B.  You could incorporate `i` into the expression being calculated for C[i] to reduce the goofiness.

This seems like a good place to reinforce the message that "A is the leader, and its parallel iterator will determine how this loop is parallelized."